File: src/main/java/com/pubnub/api/models/consumer/objects_api/PNObject.java
package com.pubnub.api.models.consumer.objects_api;

import com.google.gson.annotations.JsonAdapter;
import com.pubnub.api.models.consumer.objects_api.util.CustomPayloadJsonInterceptor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;

@Getter
@Accessors(chain = true)
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class PNObject {

    @EqualsAndHashCode.Include
    protected String id;

    @JsonAdapter(CustomPayloadJsonInterceptor.class)
    @Setter
    protected Object custom;

    protected String created;
    protected String updated;
    protected String eTag;

    protected PNObject(String id) {
        this.id = id;
    }

    protected PNObject() {
    }
}

File: src/main/kotlin/com/pubnub/api/models/consumer/objects_api/PNObject.kt
package com.pubnub.api.models.consumer.objects_api

open class PNObject(val id: String) {

    val created: String? = null
    val updated: String? = null
    val eTag: String? = null

    // todo custom interceptor
    open var custom: Any? = null

    override fun toString(): String {
        return "PNObject(" +
                "id='$id', " +
                "created=$created, " +
                "updated=$updated, " +
                "eTag=$eTag, " +
                "custom=$custom" +
                ")"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PNObject) return false

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }


}